biz.aQute.bndlib/src/aQute/bnd/osgi/MetaInfService.java

```diff
@@ -0,0 +1,295 @@
+package aQute.bnd.osgi;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import aQute.bnd.header.Attrs;
+import aQute.bnd.header.OSGiHeader;
+import aQute.bnd.header.Parameters;
+import aQute.lib.io.IO;
+import aQute.lib.strings.Strings;
+
+/**
+ * Holds our knowledge about the META-INF/services/ files.
+ * <p>
+ * These are files that are used by Java in the Service Loader. Normally they
+ * have the name of a service type and are filled with their implementations,
+ * one per line.
+ * <p>
+ * We allow these lines to be annotated like Java types. Annotations are in the
+ * comments to not disturb other users of these files. They look like:
+ *
+ * <pre>
+ * #import aQute.bnd.annotation.spi.ServiceProvider
+ * </pre>
+ */
+public class MetaInfService {
+
+	static final String		META_INF_SERVICES_STEM		= "META-INF/services";
+	static final String		META_INF_SERVICES_PREFIX	= META_INF_SERVICES_STEM + "/";
+	static final String		FQN_S						= "(\\p{javaJavaIdentifierStart}\\p{javaJavaIdentifierPart}*\\.)*\\p{javaJavaIdentifierStart}\\p{javaJavaIdentifierPart}*";
+	static final Pattern	IMPORT_P					= Pattern
+		.compile("#import\\s+(?<fqn>" + FQN_S + ")\\s*;?\\s*$");
+	static final Pattern	ANNOTATION_P				= Pattern
+		.compile("#@(?<fqn>" + FQN_S + ")\\((?<attrs>.*)" + "\\)\\s*;?\\s*$");
```

Comment:
I am ok with this, but if you change 'fqn', don't you then also not have to change the code where we get the group?

Reply:
Yes, of course the code below has to be adjusted.
In fact I mainly noticed this when reading the code below where it was first read like it would look up a fully-qualified name in the imports (which is of course not necessary), when it actually just looked up the class-name.

I just wanted to keep the suggestion short. :)